### PR TITLE
Determining props stmt link

### PR DESCRIPTION
**Pull request by FeLungs**
suggestions for the changes related to:

- Profile references within Statements
- Statement Template determining properties

### CODE REVIEW

**Comment by andyjohnson** [comment]
We have conflicting text in this PR and I wanted to make sure we are on the same page:
"Multiple references can be found in an xAPI Statement and their effect is combinatory" - this I agree with and don't think we can do it any other way

"Profile Version
A reference to a Profile Version is the broadest link between a Statement and a Profile.

{
...
"context": {"category": [{"id": "https://example.com/profile/a/version/1.0.0"}]}
...
}
This Statement MUST:

conform to either template_x or template_y "

EDIT: I realize that this example is using the logic that there are only a certain number of Statement Templates within the xAPI Profile.  We need to make that explicit, that the profiles in the examples are COMPLETE profiles and then also explain why template x or template y (e.g., they make up the complete set of templates within the xAPI profile).

**Comment by andyjohnson** [comment]
I also think we should be really explicit with how exactly Statements and conformance work.  Possibly have its own title, but otherwise looking at adding to the following text:  

"Using an introduced Concept, such as an activity type, verb, attachment usage type, extension, activity, or document resource, can be done freely, provided the defined usage and meaning are adhered to. But a Learning Record Provider can go further, and make sure to adhere to Profile-described Statement Templates and Patterns. Learning Record Providers authoring Statements SHOULD include reference(s) to a Profile by using Category Context Activities as described below."  

with "The references described in the table below is the ONLY supported method for recognizing that xAPI Statements are part of an xAPI Profile.  If a Statement does not successfully reference an xAPI Profile or one of its Statement Templates or Patterns (which effectively references that xAPI Profile), it has no recognizable relationship to any xAPI Profile or its components.  Any sort of conformance testing is not possible without those references."

**Comment by andyjohnson** [comment]
Another addition / explanation.  In the section:

"Combinations
Combining references to Profiles and Profile Components provides flexible, declarative and precise linking."

I think we should say something about not using BOTH the xAPI Profile AND the xAPI Statement Template from that xAPI Profile, as the Statement Template is sufficient.  This assumes, of course, that a validator can distinguish between the three types, which I would think it could.

**Comment by andyjohnson** [comment]
There is still a hole in the "logic" of how everything comes together.  In the examples with Profiles A and B, we are basically saying that there is some logic at play with what is placed in the context categories (for simplicity sake).  However, this isn't actually the case.  We wouldn't JUST validate a single template, we would still validate the entire profile, because that is how the algorithm works.  I think we should retool and make sure a) that the profile ID is ALWAYS included and b) that any patterns and statement template ids are only for references for assisted validation and NOT for conformance (e.g. the algorithms).

In addition, I've thought through validation and really it comes down to two philosophies - a) Statements are either validated by an external system that "knows" which profile they want to validate them against or b) Statements are validated by the xAPI Profile they reference.  In the case of "b", the xAPI Profile already references the xAPI Profile, so it being a "requirement" is inherent.  However, in "a", if the external system did indeed validate that a Statement was conformant to an xAPI Profile, the reference is lost as soon as that Statement is back "in the wild".  It should, therefore, deem NON-CONFORMANCE if it doesn't reference the xAPI Profile.  I think the following blanket rule should be applied for any profile and should be referenced as a pre-requisite to validation in "Algorithms".

"When performing tests to determine xAPI Profile conformance, all Statements MUST include the reference to the xAPI Profile they will be validated against.  To be clear, this means the xAPI Profile validation is using the same xAPI Profile (possibly write some version acceptance in here) that is referenced in the context category id property.   This could be regarded as a "special Statement Template" , but as the context category id property is not in Statement Templates, it is easier to treat it as a pre-requisite to any other validation."

**Comment by FeLungs** [comment]
> We have conflicting text in this PR and I wanted to make sure we are on the same page: "Multiple references can be found in an xAPI Statement and their effect is combinatory" - this I agree with and don't think we can do it any other way
> 
> "Profile Version A reference to a Profile Version is the broadest link between a Statement and a Profile.
> 
> { ... "context": {"category": [{"id": "https://example.com/profile/a/version/1.0.0"}]} ... } This Statement MUST:
> 
> conform to either template_x or template_y "
> 
> EDIT: I realize that this example is using the logic that there are only a certain number of Statement Templates within the xAPI Profile. We need to make that explicit, that the profiles in the examples are COMPLETE profiles and then also explain why template x or template y (e.g., they make up the complete set of templates within the xAPI profile).

On the same page and noted, I think that can be added to the "### Examples" section with the table

> I also think we should be really explicit with how exactly Statements and conformance work. Possibly have its own title, but otherwise looking at adding to the following text:
>
>"Using an introduced Concept, such as an activity type, verb, attachment usage type, extension, activity, or document resource, can be done freely, provided the defined usage and meaning are adhered to. But a Learning Record Provider can go further, and make sure to adhere to Profile-described Statement Templates and Patterns. Learning Record Providers authoring Statements SHOULD include reference(s) to a Profile by using Category Context Activities as described below."
>
>with "The references described in the table below is the ONLY supported method for recognizing that xAPI Statements are part of an xAPI Profile. If a Statement does not successfully reference an xAPI Profile or one of its Statement Templates or Patterns (which effectively references that xAPI Profile), it has no recognizable relationship to any xAPI Profile or its components. Any sort of conformance testing is not possible without those references."

If I'm understanding correctly, the new text would replace "Learning Record Providers authoring Statements SHOULD include reference(s) to a Profile by using Category Context Activities as described below." That makes sense / sounds good to me.

In terms of statement conformance to an xAPI Profile, how much of the specifics should be moved from / copied over from part 3 (profile processing libraries / algorithms)? I think the pseudocode can stay within part 3 but agreed that we need to have some kind of description / summary of how validation happens when describing the linkage between xAPI Statement and Profile.

> Another addition / explanation. In the section:
>
> "Combinations
> Combining references to Profiles and Profile Components provides flexible, declarative and precise linking."
> 
> I think we should say something about not using BOTH the xAPI Profile AND the xAPI Statement Template from that xAPI Profile, as the Statement Template is sufficient. This assumes, of course, that a validator can distinguish between the three types, which I would think it could.

Agreed. I think the most explicit way of doing this is by defining reserved Activity Types (like how the cmi5 activity type is defined within the base standard) corresponding to Profile/Template/Pattern so that there is no ambiguity at the validator or statement level.

> There is still a hole in the "logic" of how everything comes together. In the examples with Profiles A and B, we are basically saying that there is some logic at play with what is placed in the context categories (for simplicity sake). However, this isn't actually the case. We wouldn't JUST validate a single template, we would still validate the entire profile, because that is how the algorithm works. I think we should retool and make sure a) that the profile ID is ALWAYS included and b) that any patterns and statement template ids are only for references for assisted validation and NOT for conformance (e.g. the algorithms).

IMO we should expand the algorithms section to support/include single template / pattern validation (already part of the existing algorithms, just applied to each template / each pattern in a profile instead of a single, specific template / pattern)

a) I think this makes sense to include. Avoids potential profile server SPARQL query/search to find the profile and its contents
b) Assuming we expand the algorithm to include single template / pattern validation, I don't think we find ourselves in the position where the statement(s) need to be validated against the entire profile given a explicit template / pattern reference.

>In addition, I've thought through validation and really it comes down to two philosophies - a) Statements are either validated by an external system that "knows" which profile they want to validate them against or b) Statements are validated by the xAPI Profile they reference. In the case of "b", the xAPI Profile already references the xAPI Profile, so it being a "requirement" is inherent. However, in "a", if the external system did indeed validate that a Statement was conformant to an xAPI Profile, the reference is lost as soon as that Statement is back "in the wild". It should, therefore, deem NON-CONFORMANCE if it doesn't reference the xAPI Profile. I think the following blanket rule should be applied for any profile and should be referenced as a pre-requisite to validation in "Algorithms".
>
>"When performing tests to determine xAPI Profile conformance, all Statements MUST include the reference to the xAPI Profile they will be validated against. To be clear, this means the xAPI Profile validation is using the same xAPI Profile (possibly write some version acceptance in here) that is referenced in the context category id property. This could be regarded as a "special Statement Template" , but as the context category id property is not in Statement Templates, it is easier to treat it as a pre-requisite to any other validation."

Agreed. Does that establish 3 possible types of existing conformance then that should be explicitly defined? (there should probably be side discussion about naming)
 - conformant: validates against a template / pattern within profile
 - not conformant: does not validate against any template / pattern within profile
 - non conformant: does not contain profile reference and can not be validated

We may also want to split up Template and Pattern conformance sections. In the new text, what happens when a collection of statements with inconsistent profile references are provided for validation? 
- Template is straight forward, validate against templates in the profile referenced by the statement; when there's no profile reference there is no template validation. 
- Pattern has more considerations bc validation is based on multiple statements. What is the filtering logic here?
   - What happens when the Statements contain inconsistent profile references?
   - What happens when a subset of the statements don't contain a profile reference

I think the last part of the next text can be reworded along the lines of:

"...this means the xAPI Profile validation is using the xAPI Profile referenced by inScheme within the Statements category contextActivities."

I'm not following the "This could be regarded as a "special Statement Template" , but as the context category id property is not in Statement Templates, it is easier to treat it as a pre-requisite to any other validation." part. The context category id property is the Statement Templates inScheme

- profile reference pre-req, can't do validation w/o. Explanatory

File: xapi-profiles-structure.md [hunk 1]
```diff
@@ -80,7 +80,180 @@ To assist in accomplishing these two primary goals, Profiles also contain metada
 <a name="using-prof-statements"></a>
 ## 5.0 Using Profiles in Statements
 
-Using an introduced Concept, such as an activity type, verb, attachment usage type, extension, activity, or document resource, can be done freely, provided the defined usage and meaning are adhered to. But a Learning Record Provider can go further, and make sure to adhere to Profile-described Statement Templates and Patterns. Learning Record Providers authoring Statements that conform to matching Profile-described Statement Templates and Patterns SHOULD include the most up-to-date conformant Profile version as a category context activity with id equal to the version's `id` in those Statements, and Statements containing a Profile version as a category context activity MUST conform to any matching Statement Templates and Patterns that Profile version describes.
+[//]: <> (This section may need to be moved to after the Profile Component sections)
```
Comment (FeLungs) [inline xapi-profiles-structure.md:83]: adding statement validation against profile , against template, against pattern explanation needed here

- matching algorithm in natural language

Comment (FeLungs) [reply xapi-profiles-structure.md:83]: additional clarification where necessary to describe edge cases, etc.

File: xapi-profiles-structure.md [hunk 2]
```diff
@@ -80,7 +80,180 @@ To assist in accomplishing these two primary goals, Profiles also contain metada
...
+}
+```
+
+This Statement MUST:
+* conform to `template_x` or `template_y`
+* conform to `template_y` or `template_z`
+
+#### Combinations
+
+Combining references to Profiles and Profile Components provides flexible, declarative and precise linking.
```
Comment (FeLungs) [inline xapi-profiles-structure.md]: profile is sufficient for validation (assuming always included in stmt that conforms to profile)

not so much combinations but specificity, validation checks against overall template and then template / pattern references are metadata that can be utilized by xAPI Profile aware systems

-> pre-defined / reserved within xAPI Profile produced within standardization effort. Correct usage concern when testing. Do not place unnecessary restrictions on them, they are free to use like any other concept defined in a profile

**Comment by FeLungs** [comment]
Additional thought, how / should we indicate that a statement has been validated against a profile, e.g. Authority but explicit to profile verification

**Comment by FeLungs** [comment]
I'm thinking that the Profiles in Statements section and the Profile Processing Library sections should be combined into a new Part (new md doc).

That new Part could be either 3 or 4 but I'm thinking it should be part 3 because the Profile Server Part (4) will want to reference the Profile Processing Library section of the new Part.

Next commit will reflect ^^^

**Pull request merged**